src/lsptoolshost/onAutoInsertFeature.ts

```diff
@@ -0,0 +1,137 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import {
+    languages as Languages,
+    workspace as Workspace,
+    DocumentSelector as VDocumentSelector,
+    TextDocument,
+} from 'vscode';
+
+import { DynamicFeature, FeatureState, LanguageClient, RegistrationData, ensure } from 'vscode-languageclient/node';
+
+import {
+    ClientCapabilities,
+    DocumentSelector,
+    InitializeParams,
+    ProtocolRequestType,
+    RegistrationType,
+    ServerCapabilities,
+} from 'vscode-languageserver-protocol';
+
+import * as RoslynProtocol from './roslynProtocol';
+import * as Is from 'vscode-languageclient/lib/common/utils/is';
+import * as UUID from 'vscode-languageclient/lib/common/utils/uuid';
+
+export class OnAutoInsertFeature implements DynamicFeature<RoslynProtocol.OnAutoInsertRegistrationOptions> {
+    private readonly _client: LanguageClient;
+    private readonly _registrations: Map<string, RegistrationData<RoslynProtocol.OnAutoInsertRegistrationOptions>>;
+
+    constructor(client: LanguageClient) {
+        this._client = client;
+        this._registrations = new Map();
+        this.registrationType = new ProtocolRequestType<
+            RoslynProtocol.OnAutoInsertParams,
+            RoslynProtocol.OnAutoInsertResponseItem | null,
+            never,
+            void,
+            RoslynProtocol.OnAutoInsertRegistrationOptions
+        >(RoslynProtocol.OnAutoInsertRequest.method);
+    }
+    fillInitializeParams?: ((params: InitializeParams) => void) | undefined;
+    preInitialize?:
+        | ((capabilities: ServerCapabilities<any>, documentSelector: DocumentSelector | undefined) => void)
+        | undefined;
+    registrationType: RegistrationType<RoslynProtocol.OnAutoInsertRegistrationOptions>;
+    register(data: RegistrationData<RoslynProtocol.OnAutoInsertRegistrationOptions>): void {
+        if (!data.registerOptions.documentSelector) {
+            return;
+        }
+        this._registrations.set(data.id, data);
+    }
+    unregister(id: string): void {
+        const registration = this._registrations.get(id);
+        if (registration !== undefined) {
+            this._registrations.delete(id);
+        }
+    }
+    dispose(): void {
+        this._registrations.clear();
+    }
+
+    public getState(): FeatureState {
+        const selectors = this.getDocumentSelectors();
+
+        let count = 0;
+        for (const selector of selectors) {
+            count++;
+            for (const document of Workspace.textDocuments) {
+                if (Languages.match(selector, document) > 0) {
+                    return { kind: 'document', id: this.registrationType.method, registrations: true, matches: true };
+                }
+            }
+        }
+        const registrations = count > 0;
+        return { kind: 'document', id: this.registrationType.method, registrations, matches: false };
```

Comment:
I did go down that road initially. However, I wasn't able to get all the way there because it started making hardcoded assumptions about the list of language features. For example, it requires a client that implements FeatureClient<Middleware, LanguageClientOptions> but MiddleWare is hardcoded to this: https://github.com/microsoft/vscode-languageserver-node/blob/54dd4abae40251a18fbe263285e6e97f27457e62/client/src/common/client.ts#L338